File: core/python/pubplot/spec.py
"""The figure spec: one dict that both `plot` (CLI args) and `render` (config)
build, then hand to a recipe. Mirrors the R spec structure."""
import os
import yaml


def default_appearance():
    return {
        "theme": "pubplot_house",
        "palette": None,
        "x_label": None,
        "y_label": None,
        "title": None,
        "y_limits": [None, None],
        "show_points": True,
        "geom": "box",
        "bracket": {"show": True, "label": "p.signif"},
        "scale": "row",
        "cluster": "both",
    }


def spec_from_args(a):
    ap = default_appearance()
    if getattr(a, "geom", None):
        ap["geom"] = a.geom
    if getattr(a, "xlab", None):
        ap["x_label"] = a.xlab
    if getattr(a, "ylab", None):
        ap["y_label"] = a.ylab
    if getattr(a, "title", None):
        ap["title"] = a.title
    if getattr(a, "palette", None):
        ap["palette"] = [s.strip() for s in a.palette.split(",")]
    if getattr(a, "scale", None):
        ap["scale"] = a.scale
    if getattr(a, "cluster", None):
        ap["cluster"] = a.cluster
    covs = [s.strip() for s in a.covariates.split(",")] if getattr(a, "covariates", None) else None
    return {
        "engine": "python",
        "recipe": a.recipe,
        "data": {
            "file": os.path.basename(a.data), "x": a.x, "y": a.y,
            "fill": getattr(a, "fill", None), "facet": getattr(a, "facet", None),
            "time": getattr(a, "time", None), "event": getattr(a, "event", None),
            "annotation": getattr(a, "annotation", None), "covariates": covs,
        },
        "test": {
            "method": getattr(a, "test", None) or "auto",
            "paired": bool(getattr(a, "paired", False)),
            "p_adjust": getattr(a, "p_adjust", None) or "none",
        },
        "appearance": ap,
    }


def spec_from_config(path):
    cfg = yaml.safe_load(open(path))
    ap = default_appearance()
    ap.update(cfg.get("appearance") or {})
    cfg["appearance"] = ap
    t = {"method": "auto", "paired": False, "p_adjust": "none"}
    t.update(cfg.get("test") or {})
    cfg["test"] = t
    cfg.setdefault("engine", "python")
    cfg.setdefault("data", {})
    return cfg


def write_config(spec, path):
    with open(path, "w") as f:
        yaml.safe_dump(spec, f, sort_keys=False, default_flow_style=False)

File: core/r/R/spec.R
# The figure spec: one structure that both `plot` (from CLI args) and
# `render` (from plot_config.yaml) build, then hand to a recipe.

default_appearance <- function() {
  list(
    theme       = "pubplot_house",
    palette     = NULL,            # NULL means use the house palette
    x_label     = NULL,
    y_label     = NULL,
    title       = NULL,
    y_limits    = list(NULL, NULL),
    show_points = TRUE,
    geom        = "box",           # box | violin | bar | raincloud
    bracket     = list(show = TRUE, label = "p.signif"),
    scale       = "row",           # heatmap: row | column | none
    cluster     = "both",          # heatmap: both | rows | columns | none
    fc_cutoff   = 1.0,             # volcano: |log2FC| threshold
    p_cutoff    = 0.05,            # volcano: p threshold
    top_n       = 15              # volcano / enrichment: labels / terms to show
  )
}

# Build a spec from parsed command-line options.
spec_from_opt <- function(opt) {
  ap <- default_appearance()
  if (!is.null(opt$geom))    ap$geom    <- opt$geom
  if (!is.null(opt$theme))   ap$theme   <- opt$theme
  if (!is.null(opt$xlab))    ap$x_label <- opt$xlab
  if (!is.null(opt$ylab))    ap$y_label <- opt$ylab
  if (!is.null(opt$title))   ap$title   <- opt$title
  if (!is.null(opt$palette)) ap$palette <- trimws(strsplit(opt$palette, ",")[[1]])
  if (!is.null(opt$scale))     ap$scale     <- opt$scale
  if (!is.null(opt$cluster))   ap$cluster   <- opt$cluster
  if (!is.null(opt$fc_cutoff)) ap$fc_cutoff <- opt$fc_cutoff
  if (!is.null(opt$p_cutoff))  ap$p_cutoff  <- opt$p_cutoff
  if (!is.null(opt$top_n))     ap$top_n     <- opt$top_n
  list(
    engine = "r",
    recipe = opt$recipe,
    data   = list(file = basename(opt$data), x = opt$x, y = opt$y,
                  fill = opt$fill, facet = opt$facet,
                  time = opt$time, event = opt$event,
                  annotation = opt$annotation,
                  id = opt$id, group = opt$group, label = opt$label,
                  covariates = if (!is.null(opt$covariates)) {
                    trimws(strsplit(opt$covariates, ",")[[1]])
                  } else NULL),
    test   = list(
      method   = opt$test %||% "auto",
      paired   = isTRUE(opt$paired),
      p_adjust = opt$p_adjust %||% "none"
    ),
    appearance = ap
  )
}

# YAML 1.1 reads a bare `y` key as the boolean TRUE, so a hand-edited
# `y: volume` arrives as a list element named "TRUE". Map the axis keys back so
# scientists can edit plot_config.yaml naturally without quoting `y`.
fix_axis_keys <- function(d) {
  if (is.null(d) || is.null(names(d))) return(d)
  nm <- names(d)
  if ("TRUE" %in% nm && !"y" %in% nm)  nm[nm == "TRUE"]  <- "y"
  if ("FALSE" %in% nm && !"n" %in% nm) nm[nm == "FALSE"] <- "n"
  names(d) <- nm
  d
}

# Build a spec from a plot_config.yaml file, filling any missing appearance keys.
spec_from_config <- function(path) {
  cfg <- yaml::read_yaml(path)
  cfg$data <- fix_axis_keys(cfg$data)
  cfg$appearance <- modifyList(default_appearance(), cfg$appearance %||% list())
  cfg$test <- modifyList(list(method = "auto", paired = FALSE, p_adjust = "none"),
                         cfg$test %||% list())
  cfg
}

# Write a spec back out as plot_config.yaml.
write_config <- function(spec, path) {
  yaml::write_yaml(spec, path)
}
